pkg/operator/bootstrapteardown/bootstrap_teardown_controller.go

```diff
@@ -0,0 +1,300 @@
+	}
+
+	kubeAPIServer, err := c.kubeAPIServerLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+
+	kasReady := c.isKASReady(kubeAPIServer, c.configMapsGetter)
+	if !kasReady {
+		klog.Infof("Still waiting for the kube-apiserver to be ready...")
+		return fmt.Errorf("waiting for kube-apiserver to be ready")
+	}
+
+	if !c.clusterMemberShipController.IsMember("etcd-bootstrap") {
+		return nil
+	}
+
+	c.eventRecorder.Event("BootstrapTeardownController", "safe to remove bootstrap")
+	if err := c.clusterMemberShipController.RemoveBootstrap(); err != nil {
+		return err
+	}
+
+	// set bootstrap removed condition
+	_, _, updateErr := v1helpers.UpdateStatus(c.operatorConfigClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+		Type:    ConditionBootstrapRemoved,
+		Status:  operatorv1.ConditionTrue,
+		Reason:  "BootstrapNodeRemoved",
+		Message: "Etcd operator has scaled",
+	}))
+	if updateErr != nil {
+		c.eventRecorder.Warning("BootstrapTeardownErrorUpdatingStatus", updateErr.Error())
+		return updateErr
+	}
+	return nil
+}
+
+func (c *BootstrapTeardownController) isKASReady(kasOperator *operatorv1.KubeAPIServer, configMapsGetter corev1getters.ConfigMapsGetter) bool {
+	revisionMap := map[int32]struct{}{}
+	uniqueRevisions := []int32{}
+
+	for _, nodeStatus := range kasOperator.Status.NodeStatuses {
+		revision := nodeStatus.CurrentRevision
+		if _, ok := revisionMap[revision]; !ok {
+			revisionMap[revision] = struct{}{}
+			uniqueRevisions = append(uniqueRevisions, revision)
+		}
+	}
+
+	// For each revision, check that the configmap for that revision contains the
+	// appropriate storageConfig
+	for _, revision := range uniqueRevisions {
+		configMapNameWithRevision := fmt.Sprintf("%s-%d", configMapName, revision)
+		configMap, err := configMapsGetter.ConfigMaps("openshift-kube-apiserver").Get(configMapNameWithRevision, metav1.GetOptions{})
+		if err != nil {
+			klog.Errorf("doneApiServer: error getting configmap: %#v", err)
+			return false
+		}
+		if c.configMapHasRequiredValues(configMap) {
+			// if any 1 kube-apiserver pod has more than 1
+			klog.V(4).Info("kube-apiserver has required values")
+			return true
+		}
+	}
+	return false
+}
+
+type ConfigData struct {
+	StorageConfig struct {
+		Urls []string
+	}
+}
+
+func (c *BootstrapTeardownController) configMapHasRequiredValues(configMap *corev1.ConfigMap) bool {
+	config, ok := configMap.Data[configMapKey]
+	if !ok {
+		klog.V(2).Infof("configMapHasRequiredValues: config.yaml key missing configmap %s/%s", configMap.Namespace, configMap.Name)
+		c.eventRecorder.Eventf("KASconfigmapDoesNotHaveRequiredValues", "configMapHasRequiredValues: config.yaml key missing for configmap %s/%s", configMap.Namespace, configMap.Name)
+		return false
+	}
+	var configData ConfigData
+	err := json.Unmarshal([]byte(config), &configData)
+	if err != nil {
+		klog.V(2).Infof("configMapHasRequiredValues: error unmarshalling configmap %s/%s data: %#v", configMap.Namespace, configMap.Name, err)
```

Comment:
if you send an event, you don't need a separate log line.  it does tthat automatically.